src/server/routes/course.ts

```diff
@@ -0,0 +1,193 @@
+        }
+    })
+);
+
+router.delete(
+    "/delete/:course_code",
+    asyncHandler<CourseResponseType, { course_code: string }, {}>(
+        async (req, res) => {
+            const course = await Course.findOneAndDelete(req.params);
+            if (course === null) {
+                res.status(404).end();
+                return;
+            }
+            res.status(204).json(course);
+        }
+    )
+);
+
+router.patch(
+    "/update/:course_code",
+    asyncHandler<
+        CourseResponseType,
+        { course_code: string },
+        { code?: string; description?: string }
+    >(async (req, res) => {
+        const course = await Course.findOneAndUpdate(req.params, req.body);
+        if (course === null) {
+            res.status(404).end();
+            return;
+        }
+        res.status(200).json(course);
+    })
+);
+
+// TODO: Unique constraint
+router.post(
+    "/:course_code/add-activity",
+    asyncHandler<
+        CourseActivityResponseType,
+        { course_code: string },
+        CourseActivityResponseType
+    >(async (req, res) => {
+        const course = await Course.findOne(req.params);
+        if (course === null) {
+            res.status(404).end();
+            return;
+        }
+        course.activities = [...course.activities, req.body];
+        course.save();
+        res.status(200).json(req.body);
+    })
+);
+
+router.delete(
+    "/:course_code/delete-activity",
+    asyncHandler<
+        Array<CourseActivityResponseType>,
+        { course_code: string },
+        CourseActivityUnique
+    >(async (req, res) => {
+        const course = await Course.findOne(req.params);
+        if (course === null) {
+            res.status(404).end();
+            return;
+        }
+        const removed = course.activities.filter(
+            (activity) =>
+                activity.code !== req.body.code &&
+                activity.type !== req.body.type
+        );
+        course.activities = course.activities.filter(
+            (activity) =>
+                activity.code === req.body.code &&
+                activity.type === req.body.type
+        );
+        course.save();
+        res.status(200).json(removed);
+    })
+);
+
+router.get(
+    "/:course_code/activities",
+    asyncHandler<
+        Array<CourseActivityResponseType>,
+        { course_code: string },
+        CourseActivityRequestFilterType
+    >(async (req, res) => {
+        const course = await Course.findOne(req.params);
+        if (course === null) {
+            res.status(404).end();
+            return;
+        }
+        res.json(
+            course.activities.filter((activity) => {
+                const codeFilter = req.body.code || activity.code;
+                const typeFilter = req.body.type || activity.type;
+                const { chosenYear, chosenMonth } = req.body.filter;
+                let timeFilter = false;
+                // @ts-ignore
+                // TODO: type fix?
```

Comment:
You can do a for each loop e.g `activity.weeks.forEach((weekOn, i) => { ... })` to get the correct types or if you want to follow functional programming scheme you can try the `Array.reduce()` method.

Reply:
Switched to using Array.some()